jax/random.py

```diff
@@ -418,6 +419,41 @@ def _normal(key, shape, dtype):
   return onp.array(onp.sqrt(2), dtype) * lax.erf_inv(u)
 
 
+def multivariate_normal(key, mean=0.0, cov=1.0, shape=(), dtype=onp.float64):
+  dtype = xla_bridge.canonicalize_dtype(dtype)
+  return _multivariate_normal(key, mean, cov, shape, dtype)
+
+@partial(jit, static_argnums=(3, 4))
+def _multivariate_normal(key, mean, cov, shape, dtype):
+  """Sample multivariate normal random values with given shape, mean, and covariance.
+
+  Args:
+    key: a PRNGKey used as the random key.
+    mean: optional, a scalar or array of mean values along each diminsion
+    cov: optional, a scalar (isotropic), vector (diagonal covariance matrix), or full covariance matrix
+    shape: optional, a tuple of nonnegative integers representing the shape.
+
+  Returns:
+    A random array with latent dimension of (max(asarray(mean).ndim, asarray(cov).ndim)),)
+  """
+  _check_shape("multivariate_normal", shape)
+  if hasattr(cov, "shape") and cov.ndim > 0:
+      if cov.ndim > 2:
+        raise ValueError("Covariance matrix cannot have more than 2 dimensions.")
+      shape = shape + cov.shape[:1]
+      normal_samples = normal(key, shape, dtype)
+      if cov.ndim == 2:
+        samples = np.tensordot(normal_samples, cholesky(cov), axes=1)
+      else:
+        samples = normal_samples * np.sqrt(cov)
+  else:
+    if hasattr(mean, "shape") and mean.ndim > 0:
+      shape = shape + mean.shape[:1]
```

Comment:
I'm not sure I follow this bit. What's special about the first dimension of `mean`?

Reply:
The mean can either be a scalar--in which case the mean the same for each coordinate, or a full array for different means for each coordinate.


e.g. if mean = 3 for a 4 dimensional gaussian, this ultimately implies the actual mean is an array [3, 3, 3, 3]. Alternatively, specifying a mean of [1, 2, 3, 4] would center the gaussian at [1, 2, 3, 4]. 

The subtlety here is the shape of the output array is implied in part by the mean and covariance matrix shapes. If both the mean and covariance matrices scalars then line 451 will be skipped, resulting in exactly the expected behavior of a univariate gaussian.

Reply:
So a mean of 2+ dimensions should maybe be disallowed?

Reply:
Agreed. I added that check and tests to check for `ValueError`s being thrown when the covariance matrix has more than two dimensions or mean has more than one dimension.

Reply:
Thanks for the work on this!

I'm just catching up now (I'm just getting back from leave) and I think we should revise this: instead of all the special-casing (special-casing scalars, only handling 2D) I think we should only handle the vector case. People can make higher-level convenience wrappers if they want special rank polymorphism logic, but we like to keep the core libraries simple and explicit.

Since it's already merged, we can change its behavior in a follow-up.

Reply:
In other words, I think we should avoid this spec:

> The subtlety here is the shape of the output array is implied in part by the mean and covariance matrix shapes